README.md
# NAME

Graphics::ColorNames - defines RGB values for common color names

# VERSION

version v3.5.0

# SYNOPSIS

```perl
use Graphics::ColorNames;
use Graphics::ColorNames::WWW;

$pal = Graphics::ColorNames->new( qw[ X WWW ] );

$rgb = $pal->hex('green');          # returns '00ff00'
$rgb = $pal->hex('green', '0x');    # returns '0x00ff00'
$rgb = $pal->hex('green', '#');     # returns '#00ff00'

$rgb = $pal->rgb('green');          # returns '0,255,0'
@rgb = $pal->rgb('green');          # returns (0, 255, 0)
```

# DESCRIPTION

This module provides a common interface for obtaining the RGB values
of colors by standard names.  The intention is to (1) provide a common
module that authors can use with other modules to specify colors by
name; and (2) free module authors from having to "re-invent the wheel"
whenever they decide to give the users the option of specifying a
color by name rather than RGB value.

# METHODS

## `new`

The constructor is as follows:

```perl
my $pal = Graphics::ColorNames->new( @schemes );
```

where `@schemes` is an array of color schemes (palettes, dictionaries).

A valid color scheme may be the name of a color scheme (such as `X`
or a full module name such as `Graphics::ColorNames::X`), a reference
to a color scheme hash or subroutine, or to the path or open
filehandle for a `rgb.txt` file.

If none are specified, it uses the default `X` color scheme, which
corresponds to the X-Windows `rgb.txt` colors.  For most purposes,
this is good enough.  Since v3.2.0, it was updated to use the
2014-07-06 colors, so includes the standard CSS colors as well.

Other color schemes are available on CPAN,
e.g. [Graphics::ColorNames::WWW](https://metacpan.org/pod/Graphics::ColorNames::WWW).

Since version 2.1002, [Color::Library](https://metacpan.org/pod/Color::Library) dictionaries can be used as
well:

```perl
my $pal = Graphics::ColorNames->new( 'Color::Library::Dictionary::HTML' );
```

## `rgb`

```
@rgb = $pal->rgb($name);

$rgb = $pal->rgb($name, $separator);
```

If called in a list context, returns a triplet.

If called in a scalar context, returns a string separated by an
optional separator (which defauls to a comma).  For example,

```
@rgb = $pal->rgb('blue');      # returns (0, 0, 255)

$rgb = $pal->rgb('blue', ','); # returns "0,0,255"
```

Unknown color names return empty lists or strings, depending on the
context.

Color names are case insensitive, and spaces or punctuation are
ignored. So "Alice Blue" returns the same value as "aliceblue",
"ALICE-BLUE" and "a\*lICEbl-ue".  (If you are using color names based
on user input, you should add additional validation of the color
names.)

The value returned is in the six-digit hexidecimal format used in HTML and
CSS (without the initial '#'). To convert it to separate red, green, and
blue values (between 0 and 255), use the ["hex2tuple"](#hex2tuple) function.

You may also specify an absolute filename as a color scheme, if the file
is in the same format as the standard `rgb.txt` file.

## `hex`

```
$hex = $pal->hex($name, $prefix);
```

Returns a 6-digit hexidecimal RGB code for the color.  If an optional
prefix is specified, it will prefix the code with that string.  For
example,

```
$hex = $pal->hex('blue', '#'); # returns "#0000ff"
```

If the color does not exist, it will return an empty string.

A hexidecimal RGB value in the form of `#RRGGBB`, `0xRRGGBB` or
`RRGGBB` will return itself:

```
$color = $pal->hex('#123abc');         # returns '123abc'
```

## autoloaded color name methods

Autoloaded color name methods were removed in v3.4.0.

## `load_scheme`

```
$pal->load_scheme( $scheme );
```

This dynamically loads a color scheme, which can be either a hash
reference or code reference.

# EXPORTS

## `all_schemes`

```perl
my @schemes = all_schemes();
```

Returns a list of all available color schemes installed on the machine
in the `Graphics::ColorNames` namespace.

The order has no significance.

## `hex2tuple`

Converts a hexidecimal string to a tuple.

## `tuple2hex`

Converts a tuple to a hexidecimal string.

# TIED INTERFACE

The standard interface (prior to version 0.40) was through a tied hash:

```
tie %pal, 'Graphics::ColorNames', qw[ X WWW ];
```

This interface is deprecated, and will be moved to a separate module
in the future.

# CUSTOM COLOR SCHEMES

You can add naming scheme files by creating a Perl module is the name
`Graphics::ColorNames::SCHEMENAME` which has a subroutine named
`NamesRgbTable` that returns a hash of color names and RGB values.
(Schemes with a different base namespace will require the fill namespace
to be given.)

The color names must be in all lower-case, and the RGB values must be
24-bit numbers containing the red, green, and blue values in most- significant
to least- significant byte order.

An example naming schema is below:

```perl
package Graphics::ColorNames::Metallic;

sub NamesRgbTable() {
  use integer;
  return {
    copper => 0xb87333,
    gold   => 0xcd7f32,
    silver => 0xe6e8fa,
  };
}
```

You would use the above schema as follows:

```
tie %colors, 'Graphics::ColorNames', 'Metallic';
```

The behavior of specifying multiple keys with the same name is undefined
as to which one takes precedence.

As of version 2.10, case, spaces and punctuation are ignored in color
names. So a name like "Willy's Favorite Shade-of-Blue" is treated the
same as "willysfavoroteshadeofblue".  (If your scheme does not include
duplicate entrieswith spaces and punctuation, then the minimum
version of [Graphics::ColorNames](https://metacpan.org/pod/Graphics::ColorNames) should be 2.10 in your requirements.)

An example of an additional module is the [Graphics::ColorNames::Mozilla](https://metacpan.org/pod/Graphics::ColorNames::Mozilla)
module.

Since version 1.03, `NamesRgbTable` may also return a code reference:

```perl
package Graphics::ColorNames::Orange;

sub NamesRgbTable() {
  return sub {
    my $name = shift;
    return 0xffa500;
  };
}
```

See [Graphics::ColorNames::GrayScale](https://metacpan.org/pod/Graphics::ColorNames::GrayScale) for an example.

# ROADMAP

The following changes are planned in the future:

- Support for Perl versions earlier than 5.10 will be removed sometime
in 2019.
- The tied interface will be removed, but implemented in a separate
module for users that wish to use it.
- The namespace for color schemes will be moved to the
`Graphics::ColorNames::Schemes` but optios will be added to use the
existing scheme.

    This will allow modules to be named like `Graphics::ColorNames::Tied`
    without being confused for color schemes.

- This module will be rewritten to be a [Moo](https://metacpan.org/pod/Moo)-based class.

# SEE ALSO

[Color::Library](https://metacpan.org/pod/Color::Library) provides an extensive library of color schemes. A notable
difference is that it supports more complex schemes which contain additional
information about individual colors and map multiple colors to a single name.

[Color::Rgb](https://metacpan.org/pod/Color::Rgb) has a similar function to this module, but parses an
`rgb.txt` file.

[Graphics::ColorObject](https://metacpan.org/pod/Graphics::ColorObject) can convert between RGB and other color space
types.

[Graphics::ColorUtils](https://metacpan.org/pod/Graphics::ColorUtils) can also convert betweeb RGB and other color
space types, and supports RGB from names in various color schemes.

[Acme::AutoColor](https://metacpan.org/pod/Acme::AutoColor) provides subroutines corresponding to color names.

# SOURCE

The development version is on github at [https://github.com/robrwo/Graphics-ColorNames](https://github.com/robrwo/Graphics-ColorNames)
and may be cloned from [git://github.com/robrwo/Graphics-ColorNames.git](git://github.com/robrwo/Graphics-ColorNames.git)

The SourceForge project for this module at
[http://sourceforge.net/projects/colornames/](http://sourceforge.net/projects/colornames/) is no longer
maintained.

# BUGS

Please report any bugs or feature requests on the bugtracker website
[https://rt.cpan.org/Public/Dist/Display.html?Name=Graphics-ColorNames](https://rt.cpan.org/Public/Dist/Display.html?Name=Graphics-ColorNames) or
by email to
[bug-Graphics-ColorNames@rt.cpan.org](mailto:bug-Graphics-ColorNames@rt.cpan.org).

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

# AUTHOR

Robert Rothenberg <rrwo@cpan.org>

# CONTRIBUTORS

- Alan D. Salewski <alans@cji.com>
- Steve Pomeroy <xavier@cpan.org>
- "chemboy" <chemboy@perlmonk.org>
- Magnus Cedergren <magnus@mbox604.swipnet.se>
- Gary Vollink <gary@vollink.com>
- Claus Färber <cfaerber@cpan.org>
- Andreas J. König <andk@cpan.org>
- Slaven Rezić <slaven@rezic.de>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2001-2019 by Robert Rothenberg.

This is free software, licensed under:

```
The Artistic License 2.0 (GPL Compatible)
```


Changes
Revision history for {{$dist->name}}:

{{$NEXT}}

v3.5.0    2019-06-06 21:28:32+01:00 Europe/London
  [Incompatabilities]
  - Bumped minimum prerequisite to Perl v5.10.

  [Documentation]
  - Corrected typo.

  - Updated copyright year.

v3.4.0    2018-11-18 19:12:44+00:00 Europe/London
  [Incompatabilities]
  - Autoloaded color name methods were removed.

v3.3.4    2018-11-11 15:12:53+00:00 Europe/London
  [Documentation]
  - Added a ROADMAP section for planned changes.

  [Toolchain]
  - Tweaks to dist.ini.

v3.3.3    2018-10-27 19:31:14+01:00 Europe/London
  [Documentation]
  - Fix typo in GCN::X POD.

v3.3.2    2018-10-27 17:32:56+01:00 Europe/London
  [Bug Fixes]
  - Fix VERSION class method, RT 127301.

  [Toolchain]
  - Tweaks to dist.ini for simplifying build.

v3.3.1    2018-10-24 16:03:19+01:00 Europe/London
  [Tests]
  - Fixed text dependencies on modules that have been moved to
    separate distributions, RT 127430.

v3.3.0    2018-10-23 21:29:56+01:00 Europe/London
  [Incompatabilities]
  - The GCN::HTML and GCN::Windows schemes have been removed and
    released as separate modules.

  - The prefix is now ignored for hex method for invalid color names.

  [Bug Fixes]
  - Removed undef warning for unknown color names in hex and rgb methods.

  [Documentation]
  - Reorganized the documentation.

  - Documented handling of unknown color names in hex and rgb methods.

  - Added INSTALL file to the distribution.

  [Tests]
  - Reorganized and improved tests.

  [Other]
  - The tied and autoloading interfaces are deprecated.

  - Added contributor to module metadata.

v3.2.1    2018-10-06 10:59:36+01:00 Europe/London
  [Bug Fixes]
  - Fix VERSION, RT 127301.

v3.2.0    2018-10-04 00:35:01+01:00 Europe/London
  [Bug Fixes]
  - Pass through RGB hexidecimals prefixed by "0x", as claimed
    by the documentation.

  - Fix ambiguity between modiles and filenames, RT 54500.

  - Schema modules can be in any namespace.

  [Enhancements]
  - GCN::X color scheme updated to 2014-07-06 version.

  [Incompatabilities]
  - Filenames must now be absolute.

  [Tests]
  - Minor improvement in style of some tests.

v3.1.2    2018-10-01 23:14:34+01:00 Europe/London
  [Tests]
  - Removed dependency on GCN::Netscape, RT 127258.

v3.1.1    2018-10-01 17:49:09+01:00 Europe/London
  [Documentation]
  - Added note that the HTML color scheme is deprecated.

  [Bug Fixes]
  - Regexes test for ASCII digits only.

v3.1.0    2018-09-30 13:37:11+01:00 Europe/London
  [Incompatabilities]
  - The Netscape color scheme has been split out into a separate
    release.

  [Documentation]
  - Use modern syntax in some of the examples.

  - Updated contributor names.

v3.0.2    2018-09-28 17:54:49+01:00 Europe/London

  [Documentation]
  - Identify how to report bugs.

  - Add note about the old SourceForge project for this module.

  - Fixed abstract for Netscape module.

  [Other]
  - Tagged release with the RT bug database.

  - Removed Windows line endings.

v3.0.1    2018-09-28 13:39:19+01:00 Europe/London
  [Documentation]
  - Added omitted Changes to v3.0.0.

  [Tests]
  - Removed unnecessary version tests.

  [Toolchain]
  - Use newer style of module exports.

  [Other]
  - Perltidied source.

v3.0.0    2018-09-28 00:00:12+01:00 Europe/London
  [Documentation]
  - Reformated Changes file.

  - POD is built using Pod::Weaver.

  - Removed outdated sections from POD.

  - Added additional module(s) to SEE ALSO section.

  [Tests]
  - Removed QA tests that are handled by Dist::Zilla.

  [Toolchain]
  - Use Dist::Zilla to build distribution.

  [Other Changes]
  - Corrected typo in comment.

  - Cleaned up whitespace in the source code.

2.11    2008-01-10
  - updated version number for release
  - minor changes of text format
  - added fix for problems with Graphics::ColorNames::GrayScale

2.10_05 2008-01-08
  - checks if modules already loaded before re-loading them
    (uses Module::Loaded for this)
  - requires changed to load for dynamically loaded modules
  - updated README

2.10_04 2008-01-06
  - reverted back to standard object-oriented style
  - Tie::Sub is only loaded when a subroutine is used for scheme

2.10_03 2008-01-05
  - tests use Test::Exception to catch errors
  - rewrite parts of documentation to emphasize object-oriented
    interface rather than the tied interface
  - implemented a proper DESTROY method
  - misc code cleanup and reorganization

2.10_02 2008-01-04
  - added support for Color::Library dictionaries
  - minor changes to error messages
  - underscores in color names ignored
  - auto-loading color names in object-oriented interface, e.g.
    $obj->black()
  - stubs for DESTROY and UNTIE methods
  - 06-obj.t test more flexible
  - pointless 01-ColourNames test removed
  - more updates to documentation

2.10_01 2007-12-20
  - non-word characters (spaces, punctuation) now ignored in
    color names
  - smarter handling of comments in rgb.txt files
  - significant rewrite to parts of the code, requires Tie::Sub
  - rewritten how FIRSTKEY and NEXTKEY methods work
  - better error handling
  - minor updates to documentation
  - typos in comments
  - added better checking for minimum Perl version reqs
  - added aliaes w/out punctuation in Netscape scheme

2.04    2007-12-17
  - fixed inconsistencies in Netscape color scheme

2.03    2007-12-17
  - fixed glitch with version test
  - cleaned up format of text files
  - removed unneeded dependency

2.02    Sun Dec 16 2007
  - renamed developer-related test scripts
  - removed 90-prereqs.t (it wasn't working properly sometimes)
  - added DEVEL_TESTS environment variable as requirement for
    developer tests
  - README generated from POD using Pod::Readme

2.01    2007-12-16
  - Changed versions of specific color schemes to match main module
  - Removed Graphics::ColourNames from the distribution
  - updated documentation for Netscape.pm

2.0_04  2006-10-24
  - fixed Pod::Coverage tests (RT#20923)
  - various optimizations

lib/Graphics/ColorNames/HTML.pm
package Graphics::ColorNames::HTML;

# ABSTRACT: HTML color names and equivalent RGB values

=head1 SYNOPSIS

  require Graphics::ColorNames::HTML;

  $NameTable = Graphics::ColorNames::HTML->NamesRgbTable();
  $RgbBlack  = $NameTable->{black};

=head1 DESCRIPTION

This module defines color names and their associated RGB values from the
HTML 4.0 Specification.

This module is deprecated, and will be split into a separate
distribution.  You should use L<Graphics::ColorNames::WWW> instead.

=head1 KNOWN ISSUES

In versions prior to 1.1, "fuchsia" was misspelled "fuscia". This
mispelling came from un unidentified HTML specification.  It also
appears to be a common misspelling, so rather than change it, the
proper spelling was added.

=head1 SEE ALSO

L<Graphics::ColorNames>

=cut

use strict;
use warnings;

our $VERSION = 'v3.2.2';

sub NamesRgbTable() {
    use integer;
    return {
        'black'   => 0x000000,
        'blue'    => 0x0000ff,
        'aqua'    => 0x00ffff,
        'lime'    => 0x00ff00,
        'fuchsia' => 0xff00ff,    # "fuscia" is incorrect but common
        'fuscia'  => 0xff00ff,    # mis-spelling...
        'red'     => 0xff0000,
        'yellow'  => 0xffff00,
        'white'   => 0xffffff,
        'navy'    => 0x000080,
        'teal'    => 0x008080,
        'green'   => 0x008000,
        'purple'  => 0x800080,
        'maroon'  => 0x800000,
        'olive'   => 0x808000,
        'gray'    => 0x808080,
        'silver'  => 0xc0c0c0,
    };
}

1;

__END__


lib/Graphics/ColorNames/Windows.pm
package Graphics::ColorNames::Windows;

# ABSTRACT: Windows color names and equivalent RGB values

=head1 SYNOPSIS

  require Graphics::ColorNames::Windows;

  $NameTable = Graphics::ColorNames::Windows->NamesRgbTable();
  $RgbBlack  = $NameTable->{black};

=head1 DESCRIPTION

This module defines color names and their associated RGB values used in
Microsoft Windows.

=head1 SEE ALSO

C<Graphics::ColorNames>

=cut

use strict;
use warnings;

our $VERSION = 'v3.2.2';

sub NamesRgbTable() {
    use integer;
    return {
        'black'       => 0x000000,
        'blue'        => 0x0000ff,
        'cyan'        => 0x00ffff,
        'green'       => 0x00ff00,
        'magenta'     => 0xff00ff,
        'red'         => 0xff0000,
        'yellow'      => 0xffff00,
        'white'       => 0xffffff,
        'darkblue'    => 0x000080,
        'darkcyan'    => 0x008080,
        'darkgreen'   => 0x008000,
        'darkmagenta' => 0x800080,
        'darkred'     => 0x800000,
        'darkyellow'  => 0x808000,
        'darkgray'    => 0x808080,
        'lightgray'   => 0xc0c0c0,
    };
}

1;

__END__


lib/Graphics/ColorNames.pm
package Graphics::ColorNames;
use 5.006;

# ABSTRACT: defines RGB values for common color names

use strict;
use warnings;

use v5.10;

use Exporter qw/ import /;

use Carp;
use File::Spec::Functions qw/ file_name_is_absolute /;
use Module::Load 0.10;
use Module::Loaded;

our $VERSION = 'v3.5.1';

our %EXPORT_TAGS = (
    'all'     => [qw( hex2tuple tuple2hex all_schemes )],
    'utility' => [qw( hex2tuple tuple2hex )],
);
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT    = ();

sub VERSION {
    my ( $class, $wanted ) = @_;
    require version;
    return version->parse($VERSION);
}

# We store Schemes in a hash as a quick-and-dirty way to filter
# duplicates (which sometimes occur when directories are repeated in
# @INC or via symlinks).  The order does not matter.

my %FoundSchemes = ();

sub _load {
    while ( my $module = shift ) {
        unless ( is_loaded($module) ) {
            load($module);
            mark_as_loaded($module) unless ( is_loaded($module) );
        }
    }
}

# TODO - see if using Tie::Hash::Layered gives an improvement

sub _load_scheme_from_module {
    my ($self, $scheme) = @_;

    my $module =
        $scheme =~ /^\+/ ? substr( $scheme, 1 )
      : $scheme =~ /^Color::Library::Dictionary::/ ? $scheme
      :   __PACKAGE__ . '::' . $scheme;

    eval { _load($module); };
    if ($@) {
        croak "Cannot load color naming scheme module $module";
    }

    if ($module->can('NamesRgbTable')) {
        $self->load_scheme( $module->NamesRgbTable );
    }
    elsif ($module->can('_load_color_list')) {
        $self->load_scheme( $module->_load_color_list );
    }
    else {
        croak "Unknown scheme type: $module";
    }
}

sub TIEHASH {
    my $class = shift || __PACKAGE__;
    my $self = {
        _schemes  => [],
        _iterator => 0,
    };

    bless $self, $class;

    if (@_) {
        foreach my $scheme (@_) {
            if ( ref $scheme ) {
                $self->load_scheme($scheme);
            }
            elsif ($scheme =~ /^\+?(?:\w+[:][:])*\w+$/) {
                $self->_load_scheme_from_module($scheme);
            }
            elsif ( file_name_is_absolute($scheme) ) {
                $self->_load_scheme_from_file($scheme);
            }
            else {
                croak "Unknown color scheme: $scheme";
            }
        }
    }
    else {
        $self->_load_scheme_from_module('X');
    }

    return $self;
}

sub FETCH {
    my $self = shift;
    my $key = lc( shift || "" );

    # If we're passing it an RGB value, return that value

    if ( $key =~ m/^(?:\x23|0x)?([0-9a-f]{6})$/ ) {
        return $1;
    }
    else {

        $key =~ s/[^0-9a-z\%]//g;    # ignore non-word characters

        my $val = undef;
        my $i   = 0;
        while ( ( !defined $val ) && ( $i < @{ $self->{_schemes} } ) ) {
            $val = $self->{_schemes}->[ $i++ ]->{$key};
        }

        if ( defined $val ) {
            return sprintf( '%06x', $val ),;
        }
        else {
            return;
        }
    }
}

sub EXISTS {
    my ( $self, $key ) = @_;
    defined( $self->FETCH($key) );
}

sub FIRSTKEY {
    ( my $self = shift )->{_iterator} = 0;
    each %{ $self->{_schemes}->[ $self->{_iterator} ] };
}

sub NEXTKEY {
    my $self = shift;
    my ( $key, $val ) = each %{ $self->{_schemes}->[ $self->{_iterator} ] };
    unless ( defined $key ) {
        ( $key, $val ) = each %{ $self->{_schemes}->[ ++$self->{_iterator} ] };
    }
    return $key;
}

sub load_scheme {
    my $self   = shift;
    my $scheme = shift;

    if ( ref($scheme) eq "HASH" ) {
        push @{ $self->{_schemes} }, $scheme;
    }
    elsif ( ref($scheme) eq "CODE" ) {
        _load("Tie::Sub");
        push @{ $self->{_schemes} }, {};
        tie %{ $self->{_schemes}->[-1] }, 'Tie::Sub', $scheme;
    }
    elsif ( ref($scheme) eq "ARRAY" ) {

        # assumes these are Color::Library::Dictionary 0.02 files
        my $s = {};
        foreach my $rec (@$scheme) {
            my $key  = $rec->[0];
            my $name = $rec->[1];
            my $code = $rec->[5];
            $name =~ s/[\W\_]//g;    # ignore non-word characters
            $s->{$name} = $code unless ( exists $s->{$name} );
            if ( $key =~ /^(.+\:.+)\.([0-9]+)$/ ) {
                $s->{"$name$2"} = $code;
            }
        }
        push @{ $self->{_schemes} }, $s;
    }
    else {
        # TODO - use Exception
        undef $!;
        eval {
            if (   ( ref($scheme) eq 'GLOB' )
                || ref($scheme) eq "IO::File"
                || $scheme->isa('IO::File')
                || ref($scheme) eq "FileHandle"
                || $scheme->isa('FileHandle') )
            {
                $self->_load_scheme_from_file($scheme);
            }
        };
        if ($@) {
            croak "Error $@ on scheme type ", ref($scheme);
        }
        elsif ($!) {
            croak "$!";
        }
        else {
            # everything is ok?
        }
    }
}

sub _find_schemes {

    my $path = shift;

    # BUG: deep-named schemes such as Graphics::ColorNames::Foo::Bar
    # are not supported.

    if ( -d $path ) {
        my $dh = DirHandle->new($path)
          || croak "Unable to access directory $path";
        while ( defined( my $fn = $dh->read ) ) {
            if (   ( -r File::Spec->catdir( $path, $fn ) )
                && ( $fn =~ /(.+)\.pm$/ ) )
            {
                $FoundSchemes{$1}++;
            }
        }
    }
}

sub _readonly_error {
    croak "Cannot modify a read-only value";
}

sub DESTROY {
    my $self = shift;
    delete $self->{_schemes};
    delete $self->{_iterator};
}

BEGIN {
    no strict 'refs';
    *STORE  = \&_readonly_error;
    *DELETE = \&_readonly_error;
    *CLEAR  = \&_readonly_error;    # causes problems with 'undef'

    *new = \&TIEHASH;
}

# Convert 6-digit hexidecimal code (used for HTML etc.) to an array of
# RGB values

sub hex2tuple {
    my $rgb = CORE::hex(shift);
    my ( $red, $green, $blue );
    $blue  = ( $rgb & 0x0000ff );
    $green = ( $rgb & 0x00ff00 ) >> 8;
    $red   = ( $rgb & 0xff0000 ) >> 16;
    return ( $red, $green, $blue );
}

# Convert list of RGB values to 6-digit hexidecimal code (used for HTML, etc.)

sub tuple2hex {
    my ( $red, $green, $blue ) = @_;
    my $rgb = sprintf "%.2x%.2x%.2x", $red, $green, $blue;
    return $rgb;
}

sub all_schemes {
    unless (%FoundSchemes) {

        _load( "DirHandle", "File::Spec" );

        foreach my $dir (@INC) {
            _find_schemes(
                File::Spec->catdir( $dir, split( /::/, __PACKAGE__ ) ) );
        }
    }
    return ( keys %FoundSchemes );
}

sub _load_scheme_from_file {
    my $self = shift;
    my $file = shift;

    unless ( ref $file ) {
        unless ( -r $file ) {
            croak "Cannot load scheme from file: \'$file\'";
        }
        _load("IO::File");
    }

    my $fh = ref($file) ? $file : ( IO::File->new );
    unless ( ref $file ) {
        open( $fh, $file )
          || croak "Cannot open file: \'$file\'";
    }

    my $scheme = {};

    while ( my $line = <$fh> ) {
        chomp($line);
        $line =~ s/[\!\#].*$//;
        if ( $line ne "" ) {
            my $name = lc( substr( $line, 12 ) );
            $name =~ s/[\W]//g;  # remove anything that isn't a letter or number

            croak "Missing color name",
              unless ( $name ne "" );

            # TODO? Should we add an option to warn if overlapping names
            # are defined? This seems to be too common to be useful.

            # unless (exists $scheme->{$name}) {

            $scheme->{$name} = 0;
            foreach ( 0, 4, 8 ) {
                $scheme->{$name} <<= 8;
                $scheme->{$name} |= ( eval substr( $line, $_, 3 ) );
            }

            # }
        }
    }
    $self->load_scheme($scheme);

    unless ( ref $file ) {
        close $fh;
    }
}

sub hex {
    my ($self, $name, $prefix) = @_;
    my $rgb  = $self->FETCH($name);

    return '' unless defined $rgb;

    return $rgb unless defined $prefix;

    return $prefix . $rgb;
}

sub rgb {
    my ($self, $name, $separator) = @_;
    my $rgb = $self->FETCH($name);
    my @rgb = (defined $rgb) ? hex2tuple( $rgb ) : ();
    return wantarray ? @rgb : join($separator || ',', @rgb);
}

1;

__END__

=head1 SYNOPSIS

  use Graphics::ColorNames;
  use Graphics::ColorNames::WWW;

  $pal = Graphics::ColorNames->new( qw[ X WWW ] );

  $rgb = $pal->hex('green');          # returns '00ff00'
  $rgb = $pal->hex('green', '0x');    # returns '0x00ff00'
  $rgb = $pal->hex('green', '#');     # returns '#00ff00'

  $rgb = $pal->rgb('green');          # returns '0,255,0'
  @rgb = $pal->rgb('green');          # returns (0, 255, 0)

=head1 DESCRIPTION

This module provides a common interface for obtaining the RGB values
of colors by standard names.  The intention is to (1) provide a common
module that authors can use with other modules to specify colors by
name; and (2) free module authors from having to "re-invent the wheel"
whenever they decide to give the users the option of specifying a
color by name rather than RGB value.

=method C<new>

The constructor is as follows:

  my $pal = Graphics::ColorNames->new( @schemes );

where C<@schemes> is an array of color schemes (palettes, dictionaries).

A valid color scheme may be the name of a color scheme (such as C<X>
or a full module name such as C<Graphics::ColorNames::X>), a reference
to a color scheme hash or subroutine, or to the path or open
filehandle for a F<rgb.txt> file.

If none are specified, it uses the default C<X> color scheme, which
corresponds to the X-Windows F<rgb.txt> colors.  For most purposes,
this is good enough.  Since v3.2.0, it was updated to use the
2014-07-06 colors, so includes the standard CSS colors as well.

Other color schemes are available on CPAN,
e.g. L<Graphics::ColorNames::WWW>.

Since version 2.1002, L<Color::Library> dictionaries can be used as
well:

  my $pal = Graphics::ColorNames->new( 'Color::Library::Dictionary::HTML' );

=method C<rgb>

  @rgb = $pal->rgb($name);

  $rgb = $pal->rgb($name, $separator);

If called in a list context, returns a triplet.

If called in a scalar context, returns a string separated by an
optional separator (which defauls to a comma).  For example,

  @rgb = $pal->rgb('blue');      # returns (0, 0, 255)

  $rgb = $pal->rgb('blue', ','); # returns "0,0,255"

Unknown color names return empty lists or strings, depending on the
context.

Color names are case insensitive, and spaces or punctuation are
ignored. So "Alice Blue" returns the same value as "aliceblue",
"ALICE-BLUE" and "a*lICEbl-ue".  (If you are using color names based
on user input, you should add additional validation of the color
names.)

The value returned is in the six-digit hexidecimal format used in HTML and
CSS (without the initial '#'). To convert it to separate red, green, and
blue values (between 0 and 255), use the L</hex2tuple> function.

You may also specify an absolute filename as a color scheme, if the file
is in the same format as the standard F<rgb.txt> file.

=method C<hex>

  $hex = $pal->hex($name, $prefix);

Returns a 6-digit hexidecimal RGB code for the color.  If an optional
prefix is specified, it will prefix the code with that string.  For
example,

  $hex = $pal->hex('blue', '#'); # returns "#0000ff"

If the color does not exist, it will return an empty string.

A hexidecimal RGB value in the form of C<#RRGGBB>, C<0xRRGGBB> or
C<RRGGBB> will return itself:

  $color = $pal->hex('#123abc');         # returns '123abc'

=method autoloaded color name methods

Autoloaded color name methods were removed in v3.4.0.

=method C<load_scheme>

  $pal->load_scheme( $scheme );

This dynamically loads a color scheme, which can be either a hash
reference or code reference.

=export C<all_schemes>

  my @schemes = all_schemes();

Returns a list of all available color schemes installed on the machine
in the F<Graphics::ColorNames> namespace.

The order has no significance.

=export C<hex2tuple>

Converts a hexidecimal string to a tuple.

=export C<tuple2hex>

Converts a tuple to a hexidecimal string.

=head1 TIED INTERFACE

The standard interface (prior to version 0.40) was through a tied hash:

  tie %pal, 'Graphics::ColorNames', qw[ X WWW ];

This interface is deprecated, and will be moved to a separate module
in the future.

=head1 CUSTOM COLOR SCHEMES

You can add naming scheme files by creating a Perl module is the name
C<Graphics::ColorNames::SCHEMENAME> which has a subroutine named
C<NamesRgbTable> that returns a hash of color names and RGB values.
(Schemes with a different base namespace will require the fill namespace
to be given.)

The color names must be in all lower-case, and the RGB values must be
24-bit numbers containing the red, green, and blue values in most- significant
to least- significant byte order.

An example naming schema is below:

  package Graphics::ColorNames::Metallic;

  sub NamesRgbTable() {
    use integer;
    return {
      copper => 0xb87333,
      gold   => 0xcd7f32,
      silver => 0xe6e8fa,
    };
  }

You would use the above schema as follows:

  tie %colors, 'Graphics::ColorNames', 'Metallic';

The behavior of specifying multiple keys with the same name is undefined
as to which one takes precedence.

As of version 2.10, case, spaces and punctuation are ignored in color
names. So a name like "Willy's Favorite Shade-of-Blue" is treated the
same as "willysfavoroteshadeofblue".  (If your scheme does not include
duplicate entrieswith spaces and punctuation, then the minimum
version of L<Graphics::ColorNames> should be 2.10 in your requirements.)

An example of an additional module is the L<Graphics::ColorNames::Mozilla>
module.

Since version 1.03, C<NamesRgbTable> may also return a code reference:

  package Graphics::ColorNames::Orange;

  sub NamesRgbTable() {
    return sub {
      my $name = shift;
      return 0xffa500;
    };
  }

See L<Graphics::ColorNames::GrayScale> for an example.

=head1 ROADMAP

The following changes are planned in the future:

=over 4

=item *

The tied interface will be removed, but implemented in a separate
module for users that wish to use it.

=item *

The namespace for color schemes will be moved to the
C<Graphics::ColorNames::Schemes> but options will be added to use the
existing scheme.

This will allow modules to be named like C<Graphics::ColorNames::Tied>
without being confused for color schemes.

=item *

This module will be rewritten to be a L<Moo>-based class.

=back

=head1 SEE ALSO

L<Color::Library> provides an extensive library of color schemes. A notable
difference is that it supports more complex schemes which contain additional
information about individual colors and map multiple colors to a single name.

L<Color::Rgb> has a similar function to this module, but parses an
F<rgb.txt> file.

L<Graphics::ColorObject> can convert between RGB and other color space
types.

L<Graphics::ColorUtils> can also convert betweeb RGB and other color
space types, and supports RGB from names in various color schemes.

L<Acme::AutoColor> provides subroutines corresponding to color names.

=head1 append:SOURCE

The SourceForge project for this module at
L<http://sourceforge.net/projects/colornames/> is no longer
maintained.

=cut
